### PR TITLE
PERF: Cache categories in Site model take 2.

### PR DESCRIPTION
**Pull request by tgxworld**
Follow-up to aa4f0aee67d6f9802856ab4abb5a7560359854b6.

Fixed the security problem in the previous attempt.

### CODE REVIEW

File: lib/guardian/category_guardian.rb [hunk 1]
```diff
@@ -46,6 +46,11 @@ def cannot_delete_category_reason(category)
     nil
   end
 
+  def can_see_serialized_category?(category_id:, read_restricted: true)
```
Comment (tgxworld) [inline lib/guardian/category_guardian.rb:49]: Extra guard here by always defaulting to read_restricted true.

File: app/models/site.rb [hunk 2]
```diff
@@ -28,16 +28,42 @@ def user_fields
...
+      ).as_json
+    end
+  end
+
+  def categories
+    @categories ||= begin
+      categories = []
+
+      self.class.all_categories_cache.each do |category|
+        if @guardian.can_see_serialized_category?(category_id: category[:id], read_restricted: category[:read_restricted])
```
Comment (martin-brennan) [inline app/models/site.rb:63]: So we are sure these are symbols not strings :P?

File: lib/guardian/category_guardian.rb (hunk 1 above)
Comment (martin-brennan) [reply lib/guardian/category_guardian.rb:49]: Can the category[:read_restricted] on this line ever be nil or is it always true/false?

https://github.com/discourse/discourse/pull/13433/files#diff-a73755563e0f7c1dcbd6d42d4cd961ba70c6744fde3f3e8ef27c9b8d7a0e9cceR63

**Review by martin-brennan — APPROVED** [review]
I think it is fine, just a couple of questions. If you are sure this fixes the original issue it's good to go!

File: app/models/site.rb (hunk 2 above)
Comment (tgxworld) [reply app/models/site.rb:63]: Yes haha tests have been added :)

File: lib/guardian/category_guardian.rb (hunk 1 above)
Comment (tgxworld) [reply lib/guardian/category_guardian.rb:49]: It is a boolean column :)

**Pull request merged**